File: python/problems/leetcode/0208_implement_trie/solution.py
"""LeetCode 208: Implement Trie (Prefix Tree)."""

from __future__ import annotations


class Trie:
    def __init__(self) -> None:
        self._root: dict[str, dict] = {}

    def insert(self, word: str) -> None:
        cur: dict[str, dict] = self._root
        for ch in word:
            cur = cur.setdefault(ch, {})
        cur["#"] = {}

    def search(self, word: str) -> bool:
        cur = self._root
        for ch in word:
            if ch not in cur:
                return False
            cur = cur[ch]  # type: ignore[assignment]
        return "#" in cur

    def startsWith(self, prefix: str) -> bool:
        cur = self._root
        for ch in prefix:
            if ch not in cur:
                return False
            cur = cur[ch]  # type: ignore[assignment]
        return True


if __name__ == "__main__":
    t = Trie()
    t.insert("apple")
    assert t.search("apple")
    assert not t.search("app")
    assert t.startsWith("app")
    t.insert("")
    assert t.search("")
    assert t.startsWith("")
    print("0208 OK")

File: cpp/problems/leetcode/0208_implement_trie/solution.cpp
// LeetCode 208: Implement Trie (Prefix Tree)
#include <alg_std.hpp>
#include <cassert>
using namespace std;

class Trie {
    struct Node {
        array<Node*, 26> ch{};
        bool end = false;
        Node() { ch.fill(nullptr); }
    };
    Node* root;

    int idx(char c) const { return c - 'a'; }

public:
    Trie() { root = new Node(); }

    void insert(const string& word) {
        Node* cur = root;
        for (char c : word) {
            int i = idx(c);
            if (!cur->ch[i]) cur->ch[i] = new Node();
            cur = cur->ch[i];
        }
        cur->end = true;
    }

    bool search(const string& word) {
        Node* cur = root;
        for (char c : word) {
            int i = idx(c);
            if (!cur->ch[i]) return false;
            cur = cur->ch[i];
        }
        return cur->end;
    }

    bool startsWith(const string& prefix) {
        Node* cur = root;
        for (char c : prefix) {
            int i = idx(c);
            if (!cur->ch[i]) return false;
            cur = cur->ch[i];
        }
        return true;
    }
};

int main() {
    Trie t;
    t.insert("apple");
    assert(t.search("apple"));
    assert(!t.search("app"));
    assert(t.startsWith("app"));
    t.insert("");
    assert(t.search(""));
    assert(t.startsWith(""));
    cout << "0208 OK" << endl;
    return 0;
}
